subsys/zbus/zbus.c

```diff
@@ -191,21 +194,133 @@ static inline int _zbus_vded_exec(const struct zbus_channel *chan, k_timepoint_t
 	return last_error;
 }
 
+static inline void _zbus_chan_highest_priority_observer_update(const struct zbus_observer *obs)
+{
+	struct zbus_channel_observation *observation;
+
+	int count;
+
+	STRUCT_SECTION_COUNT(zbus_channel_observation, &count);
+
+	atomic_val_t obs_prio_val;
+	atomic_val_t chan_high_obs_prio_val;
+
+	for (int16_t i = 0; i < count; ++i) {
+		STRUCT_SECTION_GET(zbus_channel_observation, i, &observation);
+
+		__ASSERT(observation != NULL, "observation must be not NULL");
+
+		const struct zbus_channel *chan = observation->chan;
+
+		if (obs != observation->obs) {
+			continue;
+		}
+
+		obs_prio_val = atomic_get(&obs->data->priority);
+		chan_high_obs_prio_val = atomic_get(&chan->data->highest_observer_priority);
+
+		if ((obs_prio_val > 0) && (chan_high_obs_prio_val > obs_prio_val)) {
+			atomic_cas(&chan->data->highest_observer_priority, chan_high_obs_prio_val,
+				   obs_prio_val);
+		}
+	}
+}
+
+static void _zbus_observer_priority_set(const struct zbus_observer *obs, int8_t prio)
+{
+	__ASSERT(obs != NULL, "obs is required");
+
+	atomic_val_t obs_prio_val = atomic_get(&obs->data->priority);
+
+	if (obs_prio_val > prio) {
+		atomic_cas(&obs->data->priority, obs_prio_val, (atomic_val_t)prio);
+
+		_zbus_chan_highest_priority_observer_update(obs);
+	}
+}
+
+static int _zbus_chan_lock(const struct zbus_channel *chan, k_timeout_t timeout, int *prio)
+{
+
+	if (IS_ENABLED(CONFIG_ZBUS_PRIORITY_BOOST) && !k_is_in_isr()) {
+		int current_thread_priority = k_thread_priority_get(k_current_get());
+
+		atomic_val_t chan_high_obs_prio_val =
+			atomic_get(&chan->data->highest_observer_priority);
+
+		bool boosting = false;
+
+		if (current_thread_priority > chan_high_obs_prio_val) {
+			*prio = current_thread_priority;
+
+			LOG_DBG("Boosting %s thread priority from %d to %d", _THREAD_NAME,
+				current_thread_priority, (int)chan_high_obs_prio_val);
+
+			int p = chan_high_obs_prio_val - 1;
+
+			k_thread_priority_set(k_current_get(), MAX(p, 0));
+
```

Comment:
There is still a race condition here.  "chan_high_obs_prio_val" was read above (at line 248), and is being used here to compute "p".  But between those two lines of code another thread may have come along and changed highest_observer_priority, such that the p value you are using here is now wrong (i.e. not the highest observer priority).